src/api/encode.rs

```diff
@@ -0,0 +1,186 @@
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+
+use std::vec::IntoIter;
+use std::convert::TryFrom;
+use zmq::Message;
+
+use bitcoin::consensus::encode::{
+    serialize as consensus_serialize,
+    deserialize as consensus_deserialize
+};
+
+use crate::bp::ShortId;
+#[cfg(feature="use-rgb")]
+use crate::csv::{self, network_serialize, network_deserialize};
+use super::{Multipart, Error};
+
+
+// 1. Encoding messages
+pub trait MessageEncode where Self: Sized {
+    type Error: std::error::Error;
+    fn into_message(self) -> Message;
+    fn try_from_message(message: Message) -> Result<Self, Self::Error>;
+}
+
+/// This is a trick for rust compiler helping to distinguish types implementing
+/// mutually-exclusive traits (required until negative trait impls will be there)
+/// Implemented after concept by Martin Habovštiak <martin.habovstiak@gmail.com>
+mod strategy {
+    use core::marker::PhantomData;
+    use std::fmt;
+
+    // Defining strategies:
+    /// Strategy used for encoding data structures that support encoding with
+    /// bitcoin consensus rules (`bitcoin::consensus::encode`)
+    pub enum BitcoinConsensus { }
+
+    /// Strategy used for encoding data structures that support encoding with
+    /// RGB network serialization rules
+    #[cfg(feature="use-rgb")]
+    pub enum RGBStrategy { }
+
+    /// Strategy used for encoding data structures that can be directly
+    /// represented as `zmq::Message` with `TryFrom<Message>` and
+    /// `Into<Message>` trait implementations
+    pub enum Native { }
+
+    /// Strategy used for custom implementation of data structure encoding
+    pub trait Other { type Strategy: Clone + fmt::Debug + fmt::Display; }
+
+    pub struct Holder<T, S>(T, PhantomData<S>);
+    impl<T, S> Holder<T, S> {
+        pub fn new(val: T) -> Self { Self(val, PhantomData::<S>::default()) }
+        pub fn into_inner(self) -> T { self.0 }
+    }
+}
+
+// 1.1. Auto impl for bitcoin-serialized types
+impl<T> MessageEncode for strategy::Holder<T, strategy::BitcoinConsensus>
+    where T: bitcoin::consensus::encode::Encodable +
+             bitcoin::consensus::encode::Decodable {
+    type Error = Error;
+    fn into_message(self) -> Message {
+        Message::from(consensus_serialize(&self.into_inner()))
+    }
+    fn try_from_message(message: Message) -> Result<Self, Self::Error> {
+        Ok(Self::new(consensus_deserialize(&message)?))
+    }
+}
+
+// 1.2. Auto impl for client-validation-serialized types
+#[cfg(feature="use-rgb")]
+impl<T> MessageEncode for strategy::Holder<T, strategy::RGBStrategy>
+    where T: csv::serialize::Network {
+    type Error = Error;
```

Comment:
The general convention is to not have fixed `Error` type everywhere, but use the original error type. `type Error;` would be unnecessary otherwise, since the trait could require conversion to some specific type. `std::io::{Read, Write}` uses fixed errors (`std::io::Error`) and from my experience, it's quite unfortunate in these cases:

* Application needs to do something smart around errors (real life example: translate messages)
* `no_std` is out of question
* Inconvenient when transferring more information is needed (e.g. both OS code and file name)

So my recommendation is to prefer the original error and let the client of the library unify it if needed.